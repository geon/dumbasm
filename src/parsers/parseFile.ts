import { parseEof } from "./combinators/parseEof.js";
import { parseMonad } from "./combinators/parseMonad.js";
import { type ParserArgs, type ParseResult } from "./combinators/Parser.js";
import { parseWithErrorMessage } from "./combinators/parseWithErrorMessage.js";
import { type AsmFragment } from "./parseAsmLine.js";

export type ParsedFile = readonly AsmFragment[];

export function parseFile(...args: ParserArgs): ParseResult<ParsedFile> {
	return parseWithErrorMessage(
		"SYNTAX ERROR",
		parseMonad(parseEof, (_, { result }) => result([])),
	)(...args);
}
