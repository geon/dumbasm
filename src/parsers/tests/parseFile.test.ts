import { parseFile, type ParsedFile } from "../parseFile.js";
import { testExamples } from "../combinators/tests/testExamples.js";
import {
	createParseError,
	createParseResult,
	parsingFailed,
} from "../combinators/Parser.js";
import { suite, test, expect } from "vitest";
import { asmSamples } from "./asm-samples.js";

testExamples<ParsedFile>("parseFile", [
	{
		input: "nonsense",
		parser: parseFile,
		result: createParseError(0, "SYNTAX ERROR"),
	},
	{
		name: "empty",
		input: "",
		parser: parseFile,
		result: createParseResult(0, []),
	},
	{
		input: "tax",
		parser: parseFile,
		result: createParseResult(3, [
			{
				type: "instruction",
				value: {
					mnemonic: "tax",
					addressingMode: {
						type: "implied",
						value: undefined,
					},
				},
			},
		]),
	},
	{
		name: "\\niny\\ntax\\n\\n",
		input: "\niny\ntax\n\n",
		parser: parseFile,
		result: createParseResult(10, [
			{
				type: "instruction",
				value: {
					mnemonic: "iny",
					addressingMode: {
						type: "implied",
						value: undefined,
					},
				},
			},
			{
				type: "instruction",
				value: {
					mnemonic: "tax",
					addressingMode: {
						type: "implied",
						value: undefined,
					},
				},
			},
		]),
	},
	{
		name: "tax \\n then nonsense",
		input: "tax \n then nonsense",
		parser: parseFile,
		result: createParseError(6, "SYNTAX ERROR"),
	},
]);

suite("parseFile", () => {
	test("hello world", () => {
		const parsed = parseFile(asmSamples.helloWorld, 0);
		expect(parsingFailed(parsed)).toBe(false);
		expect(parsed).toMatchSnapshot();
	});
});
