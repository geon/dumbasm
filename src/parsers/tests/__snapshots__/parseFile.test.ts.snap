// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseFile > hello world 1`] = `
{
  "consumed": 178,
  "parsed": [
    {
      "type": "instruction",
      "value": {
        "addressingMode": {
          "type": "immediate",
          "value": 0,
        },
        "mnemonic": "ldy",
      },
    },
    {
      "type": "label",
      "value": "loop",
    },
    {
      "type": "instruction",
      "value": {
        "addressingMode": {
          "type": "indexed",
          "value": {
            "index": "Y",
            "operand": {
              "type": "identifier",
              "value": "message",
            },
          },
        },
        "mnemonic": "lda",
      },
    },
    {
      "type": "instruction",
      "value": {
        "addressingMode": {
          "type": "address",
          "value": {
            "type": "identifier",
            "value": "done",
          },
        },
        "mnemonic": "beq",
      },
    },
    {
      "type": "instruction",
      "value": {
        "addressingMode": {
          "type": "address",
          "value": {
            "type": "number",
            "value": 65490,
          },
        },
        "mnemonic": "jsr",
      },
    },
    {
      "type": "instruction",
      "value": {
        "addressingMode": {
          "type": "implied",
          "value": undefined,
        },
        "mnemonic": "iny",
      },
    },
    {
      "type": "instruction",
      "value": {
        "addressingMode": {
          "type": "address",
          "value": {
            "type": "identifier",
            "value": "loop",
          },
        },
        "mnemonic": "jmp",
      },
    },
    {
      "type": "label",
      "value": "done",
    },
    {
      "type": "instruction",
      "value": {
        "addressingMode": {
          "type": "implied",
          "value": undefined,
        },
        "mnemonic": "rts",
      },
    },
    {
      "type": "label",
      "value": "message",
    },
    {
      "type": "directive",
      "value": ".byte 147 ; clear",
    },
    {
      "type": "directive",
      "value": ".text "HELLO WORLD"",
    },
    {
      "type": "directive",
      "value": ".byte 0 ; null terminator",
    },
  ],
  "type": "result",
}
`;
